### PR TITLE
New bonus guide: LNDg

### LINKED ISSUE #1125: LNDG is not shown in browser with 502 bad Gateway.
I followed latest updated Lndg guide_10/13 (there is a link below). Before installing uwsgi, it worked on http://raspibolt.local:8889 but After setting up uwsgi, nginx, etc (web server configuration part), it no longer works on https://raspibolt.local:8889. It says 502 bad gateway. 
I restarted uwsgi, nginx but didn't work. And nginx error log file (/var/log/nginx/error.log) is as following :

2022/10/14 03:37:16 [crit] 990#990: *202 connect() to unix:///home/lndg/lndg/lndg.sock failed (13: Permission denied) while connecting to upstream, client: xx.xx.xx.xx, server: _, request: "GET / HTTP/1.1", upstream: "uwsgi://unix:///home/lndg/lndg/lndg.sock:", host: "raspibolt.local:8889"

lndg.sock file has been applied 660 well by the chmod, chown commands. 
srw-rw---- 1 lndg www-data     0 Oct 18 02:14 lndg.sock
I also restarted nginx, uwsgi but still don't work. 
I dont' know what's going on. Will appreciate with your help. 

### Location in guide
https://github.com/VajraOfIndra/RaspiBolt/blob/LNDg/guide/bonus/lightning/lndg.md

### Environment
 - Hardware platform: PC (amd64)
 - Operating system: ubuntu 22.04

### PR DESCRIPTION
**Pull request by VajraOfIndra**
#### What

LNDg is a lite GUI web interface to help you manually manage your node and automate operations such as rebalancing, fee adjustemnts and channel node opening.

Has been requested by the community: e.g. https://t.me/raspibolt/12058

#### How

A new bonus guide in the +Lightning section

#### Scope

- [ ] significant change to core configuration
- [x] independent bonus guide
- [ ] simple bug fix

Fixes #1125

#### Test & maintenance

How can the changes be tested? Is there ongoing maintenance effort? If this is a bonus guide: are you willing to update it from time to time?

#### Animated GIF (optional)

hammock

### CODE REVIEW

**Comment by twofaktor** [comment]
Where is the `Remote Tor connection`?

**Comment by VajraOfIndra** [comment]
> Where is the `Remote access over Tor` section? Is not necessary?

It's not necessary but would be nice to have for sure. You could make a PR in my repo if you want, or else someone could add it later once it's merged. I let you manage that as i wont have time to add the Tor sectio myself.

File: guide/bonus/lightning/lndg.md [hunk 1]
```diff
@@ -0,0 +1,662 @@
...
+### HTLC failure stream data
+
+* Log in with user "lndg" and enter the LNDg repository
+
+  ```sh
+  $ sudo su - lndg
+  $ cd lndg
+  ```
+
+* As user “lndg”, create a bash file to call `rebalance.py` and paste the following lines. Save and exit.
```
Comment (martinatime) [inline guide/bonus/lightning/lndg.md:575]: I think this is the wrong python script for this section

File: guide/bonus/lightning/lndg.md [hunk 2]
```diff
@@ -0,0 +1,662 @@
...
+  /home/lndg/lndg/.venv/bin/python /home/lndg/lndg/htlc_stream.py
+  ```
+* Make the script executable by user “lndg” then exit the "lndg" user session
+
+  ```sh
+  $ chmod u+x htlc_stream.sh
+  $ exit
+  ```
+
+* Create a service file for the `rebalancer.sh` script
```
Comment (martinatime) [inline guide/bonus/lightning/lndg.md:593]: Wrong shell script for this section

File: guide/bonus/lightning/lndg.md [hunk 3]
```diff
@@ -0,0 +1,662 @@
...
+
+  ```sh
+  $ sudo systemctl enable htlc-stream-lndg.service
+  $ sudo systemctl start htlc-stream-lndg.service
+  $ sudo systemctl status htlc-stream-lndg.service
+  ```
+
+---
+
+## For the future: LNBits update
```
Comment (martinatime) [inline guide/bonus/lightning/lndg.md:625]: LNBits? Do you mean LNDg?

**Comment by martinatime** [comment]
I think this guide is probably in mostly good shape. I left some comments where I think there was some copy/paste errors.

File: guide/bonus/lightning/lndg.md (hunk 3 above)
Comment (VajraOfIndra) [reply guide/bonus/lightning/lndg.md]: Fixed! Thanks

File: guide/bonus/lightning/lndg.md (hunk 2 above)
Comment (VajraOfIndra) [reply guide/bonus/lightning/lndg.md]: Fixed (I think I got them all, PTAL)

File: guide/bonus/lightning/lndg.md (hunk 1 above)
Comment (VajraOfIndra) [reply guide/bonus/lightning/lndg.md]: fixed

**Comment by VajraOfIndra** [comment]
> I think this guide is probably in mostly good shape. I left some comments where I think there was some copy/paste errors.

I have some issue with the uwsgi/nginx section. It comes a pull request I merged without testing myself at the time. I think I might remove that section for the time being. We could always add it later via a tested PR once it works. Same with remote connection through Tor.` Wdyat?

File: guide/bonus/lightning/lndg.md [hunk 4]
```diff
@@ -0,0 +1,783 @@
...
+
+  ```sh
+  $ sudo nano /etc/systemd/system/uwsgi.service
+  ```
+  
+  ```ini
+  [Unit]
+  Description=LNDg uWSGI app
+  After=lnd.service
+  PartsOf=lndg.service
```
Comment (blckbx) [inline guide/bonus/lightning/lndg.md:237]:
```suggestion
  PartOf=lnd.service
```

**Comment by VajraOfIndra** [comment]
I reviewd the guide and fixed a few major issues:
- first start URL is clearnet http (not https)
- removed gloabl apt install uwsgi (is installed via Python pip)
- Removed outdated StandardError of service file
- made the backup services (jobs, rebalancer etc) dependent on lnd.service (suing After and PartOf). We could make them dependent on uwsgi.service maybe?
The guide works on my node.

**Comment by VajraOfIndra** [comment]
I simplified the guide by making the backend services and timers dependent on `uwsgi.service`, which is itself dependent on `lnd.service`.

Like this, to stop or start LNDg and the backend services (e.g., when updating), we'll just need to start and stop `uwsgi.service` (all the other timers and services will start and stop automatically). And when we start and stop LND, LNDg/`uwsgi.service` will start and stop as well automatically.

|Service|`After`|`PartOf`|
|-------|------|-------|
|`uwsgi.service`|`lnd.service`|`lnd.service`|
|`jobs-lndg.service`|||
|`jobs-lndg.timer`|`uwsgi.service`|`uwsgi.service`|
|`rebalancer-lndg.service`|||
|`rebalancer-lndg.timer`|`uwsgi.service`|`uwsgi.service`|
|`htlc-stream-lndg.service`|`uwsgi.service`|`uwsgi.service`|

File: guide/bonus/lightning/lndg.md [hunk 5]
```diff
@@ -0,0 +1,796 @@
...
+
+We do not want to run LNDg alongside `bitcoind` and `lnd` because of security reasons. 
+For that we will create a separate user and we will be running the code as the new user. 
+
+* Create a new user and make it a member of the "lnd" group to give it read access to the LND macaroons and data
+  
+  ```sh
+  $ sudo adduser --disabled-password --gecos "" lndg
+  $ sudo adduser lndg lnd
+  $ sudo adduser lndg www-data
```
Comment (blckbx) [inline guide/bonus/lightning/lndg.md:81]: In the end, I also got `502 gateway timeout` as mentioned in #1125. Solution was to add `www-data` to `lndg` group:
```sh
$ sudo adduser www-data lndg
```

File: guide/bonus/lightning/lndg.md [hunk 6]
```diff
@@ -0,0 +1,796 @@
...
+  RestartSec=30
+  KillSignal=SIGQUIT
+  Type=notify
+  NotifyAccess=all
+  
+  [Install]
+  WantedBy=sockets.target
+  ```
+
+* Create a nginx configuration file for the LNDg website with a server listening on port 4890
```
Comment (blckbx) [inline guide/bonus/lightning/lndg.md:250]: The listening port is set to 8889 in config below.

File: guide/bonus/lightning/lndg.md [hunk 7]
```diff
@@ -0,0 +1,796 @@
...
+
+---
+
+## Dashboard privacy configuration
+
+LNDg offers the possibility to create links to blockchain and lightning explorers on node aliases and transaction IDs. By default, LNDg uses public websites 1ml.com for its lightning explorer and mempool.space for its blockchain explorer.
+
+### Blockchain explorer
+
+To preserve privacy it is better that you use your own self-hosted blockhain explorer (e.g. the BTC RPC Explorer.
```
Comment (blckbx) [inline guide/bonus/lightning/lndg.md:637]:
```suggestion
To preserve privacy it is better that you use your own self-hosted blockhain explorer (e.g. the BTC RPC Explorer).
```

File: guide/bonus/lightning/lndg.md [hunk 8]
```diff
@@ -0,0 +1,796 @@
...
+
+## Dashboard privacy configuration
+
+LNDg offers the possibility to create links to blockchain and lightning explorers on node aliases and transaction IDs. By default, LNDg uses public websites 1ml.com for its lightning explorer and mempool.space for its blockchain explorer.
+
+### Blockchain explorer
+
+To preserve privacy it is better that you use your own self-hosted blockhain explorer (e.g. the BTC RPC Explorer. 
+
+* Open your LNDg website at https://raspibolt.local:8889 (replace raspibolt.local by your nod's IP address if necessary)
```
Comment (blckbx) [inline guide/bonus/lightning/lndg.md:639]:
```suggestion
* Open your LNDg website at https://raspibolt.local:8889 (replace raspibolt.local by your node's IP address if necessary)
```

**Review by blckbx — COMMENTED** [review]
some nits, one bugfix, rest works like a charm! 🥳

File: guide/bonus/lightning/lndg.md (hunk 5 above)
Comment (VajraOfIndra) [reply guide/bonus/lightning/lndg.md:81]: I added that step

**Review by blckbx — APPROVED** [review]
tACK (amd64)

**Comment by VajraOfIndra** [comment]
> tACK (amd64)

Thanks for the review. Is it ready to merge you think?

Btw, just made a little [change ](https://github.com/raspibolt/raspibolt/pull/1070/commits/d6d3d60a7e7e906a9b161e3c8fa9ce2bcba4b172)to make the backend service and timers start when uwsgi starts (as PartOf does not take care of that). 

To test it:
* Add the new line
* Register the change
  `$ sudo systemctl daemon-reload`

* Stop LNDg
  `$ sudo systemctl stop uwsgi.service`

* Check that `jobs-lndg.timer` has stopped (triggered by the `PartOf` option in `jobs-lndg.timer`)
  ```
  $ sudo systemctl status jobs-lndg.timer
  > [...]
  >  Active: inactive (dead)
  ```

* Starts LNDg
  `$ sudo systemctl start uwsgi.service`

* Check that `jobs-lndg.timer` started automatically (trigerred by the `Wants` option in `uwsgi.service`)
  ```
  $ sudo systemctl status jobs-lndg.timer
  > [...]
  >  Active: active (running)
  ```

**Comment by blckbx** [comment]
> Thanks for the review. Is it ready to merge you think?

Regarding the last change:  I already updated my local installation and it works perfectly fine.

I think the PR is in very good shape, it's probably a good idea to wait for feedback on your TG question that could strip down the guide quite a bit: https://t.me/c/1442363505/33435

Other than that: another thing mentioned in LNDg group and also done in RaspiBlitz' PR: they moved `/home/lndg/lndg/data` directory to `/data/lndg/data` and symlinked it. wdyt about that?

**Comment by VajraOfIndra** [comment]
> Regarding the last change: I already updated my local installation and it works perfectly fine.

Good to hear!

> I think the PR is in very good shape, it's probably a good idea to wait for feedback on your TG question that could strip down the guide quite a bit: https://t.me/c/1442363505/33435

Good idea, will wait to hear from NGU

> Other than that: another thing mentioned in LNDg group and also done in RaspiBlitz' PR: they moved `/home/lndg/lndg/data` directory to `/data/lndg/data` and symlinked it. wdyt about that?

This seems to make sense indeed. Nice catch.
What goes in the LNDg data folder? some sort of database?

**Comment by blckbx** [comment]
> This seems to make sense indeed. Nice catch. What goes in the LNDg data folder? some sort of database?

Yeah, settings, stats, everything worth being stored.

**Comment by VajraOfIndra** [comment]
> Other than that: another thing mentioned in LNDg group and also done in RaspiBlitz' PR: they moved `/home/lndg/lndg/data` directory to `/data/lndg/data` and symlinked it. wdyt about that?

I added a section to do this:
* used rsync instead of cp to avoid corruption of DB
* stored the DB in `/data/lndg` rather than `/data/lndg/data` as extra directory not necessary
* new section is done after the step that removes the password file as this file is in the same data directory as the DB

PTAL :)

File: guide/bonus/lightning/lndg.md [hunk 9]
```diff
@@ -144,6 +144,37 @@ If you didn't save the password, you can get it again with: `nano /home/lndg/lnd
...
+  $ ls -la /data/lndg
+  > drwxr-xr-x  2 lndg    lndg        4096 Nov 11 11:24 .
+  > drwxr-xr-x 14 admin   admin       4096 Nov 11 11:19 ..
+  > -rw-r--r--  1 lndg    lndg    19468288 Nov 11 11:23 db.sqlite3
+  ```
+
+* Delete the old database from the LNDg runtime directory and instead create a symbolic link 
+
+  ```sh
+  $ sudo rm /home/lndg/lndg/data/db.sqlite
```
Comment (blckbx) [inline guide/bonus/lightning/lndg.md:171]:
```suggestion
  $ sudo rm /home/lndg/lndg/data/db.sqlite3
```

**Review by blckbx — COMMENTED** [review]
bad filename

**Comment by blckbx** [comment]
tACK²

**Pull request merged**